iceoryx_utils/test/moduletests/test_concurrent_smart_lock.cpp

```diff
@@ -0,0 +1,358 @@
+    SmartLockTester()
+    {
+    }
+
+    SmartLockTester(const int32_t a)
+        : m_a(a)
+    {
+    }
+
+    SmartLockTester(const SmartLockTester& rhs)
+        : m_a(rhs.m_a)
+    {
+        rhs.m_b = rhs.m_b + 1;
+    }
+
+    SmartLockTester(SmartLockTester&& rhs)
+        : m_a(rhs.m_a)
+    {
+        rhs.m_a = 0;
+        rhs.m_b = rhs.m_b + 1;
+    }
+
+    SmartLockTester& operator=(const SmartLockTester& rhs)
+    {
+        if (this != &rhs)
+        {
+            rhs.m_b = rhs.m_b + 1;
+            m_a = rhs.m_a;
+        }
+        return *this;
+    }
+
+    SmartLockTester& operator=(SmartLockTester&& rhs)
+    {
+        if (this != &rhs)
+        {
+            rhs.m_b = rhs.m_b + 1;
+            m_a = rhs.m_a;
+            rhs.m_a = 0;
+        }
+        return *this;
+    }
+
+    ~SmartLockTester()
+    {
+    }
+
+    int32_t getA() const
+    {
+        return m_a;
+    }
+
+    int32_t getB() const
+    {
+        return m_b;
+    }
+
+    void setA(const int32_t a)
+    {
+        m_a = a;
+    }
+
+    void incrementA()
+    {
+        m_a = m_a + 1;
+    }
+
+    void constIncrementA() const
+    {
+        m_a = m_a + 1;
+    }
+
+  private:
+    mutable int32_t m_a = 0;
+    mutable int32_t m_b = 0;
+};
+
+class smart_lock_test : public Test
+{
+  public:
+    void SetUp() override
+    {
+    }
+    void TearDown() override
+    {
+    }
+
+    void waitUntilThreadsHaveStarted(const uint64_t numberOfThreads)
+    {
+        ++m_numberOfThreadWaiter;
+        while (m_numberOfThreadWaiter.load() != numberOfThreads)
+        {
+        }
+    }
+
+    using SutType_t = smart_lock<SmartLockTester>;
+    optional<SutType_t> m_sut;
+    std::atomic<uint64_t> m_numberOfThreadWaiter{0U};
+};
+constexpr uint64_t NUMBER_OF_RUNS_PER_THREAD = 100000U;
+constexpr uint64_t NUMBER_OF_THREADS = 4;
+
+} // namespace
+
+//////////////////////////////////////
+// BEGIN single threaded api test
+//////////////////////////////////////
+
+TEST_F(smart_lock_test, defaultConstructionOfUnderlyingObjectWorks)
+{
+    m_sut.emplace();
+    EXPECT_THAT((*m_sut)->getA(), Eq(0));
+}
+
+TEST_F(smart_lock_test, constructionWithOneValueCTorOfUnderlyingObjectWorks)
+{
+    constexpr int32_t CTOR_VALUE = 25;
+    m_sut.emplace(ForwardArgsToCTor, CTOR_VALUE);
+    EXPECT_THAT((*m_sut)->getA(), Eq(CTOR_VALUE));
+}
+
+TEST_F(smart_lock_test, copyConstructionOfUnderlyinObjectWorks)
+{
+    constexpr int32_t CTOR_VALUE = 121;
+    SmartLockTester tester(CTOR_VALUE);
+    m_sut.emplace(ForwardArgsToCTor, tester);
+    EXPECT_THAT((*m_sut)->getA(), Eq(CTOR_VALUE));
+    EXPECT_THAT(tester.getA(), Eq(CTOR_VALUE));
+}
+
+TEST_F(smart_lock_test, moveConstructionOfUnderlyinObjectWorks)
+{
+    constexpr int32_t CTOR_VALUE = 1211;
+    SmartLockTester tester(CTOR_VALUE);
+    m_sut.emplace(ForwardArgsToCTor, std::move(tester));
+    EXPECT_THAT((*m_sut)->getA(), Eq(CTOR_VALUE));
+    EXPECT_THAT(tester.getA(), Eq(0));
```

Comment:
instead of having this implicit assumption that `a == 0 == moved`, could you add an explicit `isMoved` method with an explicit `m_isMoved` member or something similar. This makes maintaining the tests simpler